File: shared/clock/FakeClock.ts
import type { IClock } from "./IClock";

/**
 * Controllable clock for tests. Lets deadline/timeout tests advance time in a single
 * call instead of actually sleeping, so a 45-second scenario runs in milliseconds.
 */
export class FakeClock implements IClock {
  private currentMs: number;

  constructor(startAtMs = 0) {
    this.currentMs = startAtMs;
  }

  now(): number {
    return this.currentMs;
  }

  advanceBy(ms: number): void {
    if (ms < 0) {
      throw new RangeError("FakeClock cannot advance by a negative duration");
    }
    this.currentMs += ms;
  }

  setTo(ms: number): void {
    this.currentMs = ms;
  }
}

File: src/Gweb.Shared/Clock/FakeClock.cs
namespace Gweb.Shared.Clock;

/// <summary>
/// Controllable clock for tests. Lets deadline/timeout tests advance time in a single
/// call instead of actually sleeping, so a 45-second scenario runs in milliseconds.
/// </summary>
public sealed class FakeClock(long startAtMs = 0) : IClock
{
    private long _currentMs = startAtMs;

    public long NowMs() => _currentMs;

    public void AdvanceBy(long ms)
    {
        if (ms < 0)
        {
            throw new ArgumentOutOfRangeException(nameof(ms), ms, "FakeClock cannot advance by a negative duration");
        }
        _currentMs += ms;
    }

    public void SetTo(long ms) => _currentMs = ms;
}
